### PR TITLE
expression: add builtin function `WEIGHT_STRING()`

### LINKED ISSUE #3580: builtin function WEIGHT_STRING() not support
Please answer these questions before submitting your issue. Thanks!

1. What did you do?

     use mysql builtin function WEIGHT_STRING(), but it is not supported by TiDB yet.

2. What did you expect to see?

```
mysql> SELECT HEX(WEIGHT_STRING('ab' AS CHAR(4)));
+-------------------------------------+
| HEX(WEIGHT_STRING('ab' AS CHAR(4))) |
+-------------------------------------+
| 0041004200200020                    |
+-------------------------------------+
1 row in set (0.03 sec)
```

3. What did you see instead?

```
mysql> SELECT HEX(WEIGHT_STRING('ab' AS CHAR(4)));
ERROR 1105 (HY000): line 0 column 25 near "'ab' AS CHAR(4)))" (total length 42)
```

4. What version of TiDB are you using (`tidb-server -V`)?

```
[wink@localhost tidb]$ bin/tidb-server -V
Git Commit Hash: 677fd05c82517e20c60345202e8dc612b1942615
UTC Build Time:  2017-06-29 08:03:08
```

### PR DESCRIPTION
**Pull request by bb7133**
### What problem does this PR solve? 
This PR tries to close #3580 

Related parser PR: https://github.com/pingcap/parser/pull/743

### What is changed and how it works?
Add builtin-function `WEIGHT_STRING()`. [MySQL manual](https://dev.mysql.com/doc/refman/8.0/en/string-functions.html#function_weight-string) described most of its behavior except:

- If the input `str` is of numeric types, `WEIGHT_STRING()` in MySQL returns `NULL`, so does this implementation.
- If the input `str` is of other types rather than string or numeric, the result of `WEIGHT_STRING()` in MySQL is not cleared. I simply evaluated the input as string. 

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change

Side effects

 - NA

### CODE REVIEW

**Comment by bb7133** [comment]
PTAL @wjhuang2016 @qw4990

File: expression/builtin_string.go [hunk 1]
```diff
@@ -3627,3 +3630,154 @@ type loadFileFunctionClass struct {
...
+	str, isNull, err := b.args[0].EvalString(b.ctx, row)
+	if err != nil {
+		return "", true, err
+	}
+	if isNull {
+		return "", true, nil
+	}
+
+	// TODO: refactor padding codes after padding is supported by collators.
+	switch b.padding {
```
Comment (wjhuang2016) [inline expression/builtin_string.go]: Please update it since the padding PR is merged.

Comment (bb7133) [reply expression/builtin_string.go]: Done.

File: expression/builtin_string.go [hunk 2]
```diff
@@ -3627,3 +3630,158 @@ type loadFileFunctionClass struct {
...
+
+const (
+	// weightStringPaddingNone is used for WEIGHT_STRING(expr) if the expr is non-numeric.
+	weightStringPaddingNone weightStringPadding = 0xFF
+	// weightStringPaddingAsChar is used for WEIGHT_STRING(expr AS CHAR(x)) and the expr is non-numeric.
+	weightStringPaddingAsChar = 0x20
+	// weightStringPaddingAsBinary is used for WEIGHT_STRING(expr as BINARY(x)) and the expr is non-numeric.
+	weightStringPaddingAsBinary = 0x00
+	// weightStringPaddingNull is used for WEIGHT_STRING(expr [AS (CHAR|BINARY)]) if the expr is numeric.
+	weightStringPaddingNull = 0xFE
```
Comment (kennytm) [inline expression/builtin_string.go]: On MySQL 8.0 this "NULL" transformation only applies to `CHAR`.

```sql
mysql> select weight_string(456789 as binary(4)), weight_string(3.1415926535e0 as binary(4)), weight_string(456789 as char(4)), weight_string(3.1415926535e0 as char(4))\G
*************************** 1. row ***************************
        weight_string(456789 as binary(4)): 4567
weight_string(3.1415926535e0 as binary(4)): 3.14
          weight_string(456789 as char(4)): NULL
  weight_string(3.1415926535e0 as char(4)): NULL
1 row in set, 2 warnings (0.00 sec)

mysql> show warnings;
+---------+------+-----------------------------------------------------+
| Level   | Code | Message                                             |
+---------+------+-----------------------------------------------------+
| Warning | 1292 | Truncated incorrect BINARY(4) value: '456789'       |
| Warning | 1292 | Truncated incorrect BINARY(4) value: '3.1415926535' |
+---------+------+-----------------------------------------------------+
2 rows in set (0.00 sec)
```

(The same for MariaDB.)

Comment (bb7133) [reply expression/builtin_string.go]: Addressed, thanks!

**Comment by bb7133** [comment]
PTAL @qw4990 @kennytm @wjhuang2016

File: expression/builtin_string.go [hunk 3]
```diff
@@ -3627,3 +3630,163 @@ type loadFileFunctionClass struct {
...
+	baseBuiltinFunc
+
+	padding weightStringPadding
+	length  int
+}
+
+func (b *builtinWeightStringSig) Clone() builtinFunc {
+	newSig := &builtinWeightStringSig{}
+	newSig.cloneFrom(&b.baseBuiltinFunc)
+	newSig.length = b.length
```
Comment (qw4990) [inline expression/builtin_string.go]: newSig.padding = b.padding?

File: expression/builtin_string_test.go [hunk 4]
```diff
@@ -2306,3 +2306,72 @@ func (s *testEvaluatorSuite) TestStringRight(c *C) {
...
+		c.Assert(err, IsNil)
+		// Reset warnings.
+		s.ctx.GetSessionVars().StmtCtx.ResetForRetry()
+		result, err := evalBuiltinFunc(f, chunk.Row{})
+		c.Assert(err, IsNil)
+		if result.IsNull() {
+			c.Assert(test.expect, IsNil)
+			continue
+		}
+		res, err := result.ToString()
```
Comment (wjhuang2016) [inline expression/builtin_string_test.go]: ToString may be wrong if the collation is general_ci

File: expression/builtin_string.go [hunk 5]
```diff
@@ -3627,3 +3630,163 @@ type loadFileFunctionClass struct {
 func (c *loadFileFunctionClass) getFunction(ctx sessionctx.Context, args []Expression) (builtinFunc, error) {
 	return nil, errFunctionNotExists.GenWithStackByArgs("FUNCTION", "load_file")
 }
+
+type weightStringPadding byte
+
+const (
+	// weightStringPaddingNone is used for WEIGHT_STRING(expr) if the expr is non-numeric.
+	weightStringPaddingNone weightStringPadding = 0xFF
```
[the remark this answers is not in the document]
Comment (bb7133) [reply expression/builtin_string.go]: The value means nothing really, it's just an enum I picked.

File: expression/builtin_string_test.go (hunk 4 above)
Comment (bb7133) [reply expression/builtin_string_test.go]: That's fine, the `string` in Go does not have to be a valid `utf8` sequence.

File: expression/builtin_string.go (hunk 3 above)
Comment (bb7133) [reply expression/builtin_string.go]: Done, thanks.

File: expression/builtin_string_test.go (hunk 4 above)
Comment (bb7133) [reply expression/builtin_string_test.go]: I've updated the tests for `utf8mb4_general_ci`.

**Comment by bb7133** [comment]
Addressed, PTAL @wjhuang2016 @qw4990

File: expression/builtin_string.go [hunk 6]
```diff
@@ -3627,3 +3630,164 @@ type loadFileFunctionClass struct {
...
+	weightStringPaddingAsBinary = 0x00
+	// weightStringPaddingNull is used for WEIGHT_STRING(expr [AS (CHAR|BINARY)]) for all other cases, it returns null always.
+	weightStringPaddingNull = 0xFE
+)
+
+type weightStringFunctionClass struct {
+	baseFunctionClass
+}
+
+func newWeightStringFunctionClass() *weightStringFunctionClass {
```
Comment (djshow832) [inline expression/builtin_string.go]: This function is never used, and other `functionClass` didn't define a `new` function.

File: expression/builtin_string_test.go [hunk 7]
```diff
@@ -2306,3 +2308,123 @@ func (s *testEvaluatorSuite) TestStringRight(c *C) {
...
+		{nil, "BINARY", 5, nil},
+		{7, "BINARY", 2, "7\x00"},
+		{"a", "BINARY", 1, "a"},
+		{"ab", "BINARY", 1, "a"},
+		{"a", "BINARY", 5, "a\x00\x00\x00\x00"},
+		{"a ", "BINARY", 5, "a \x00\x00\x00"},
+		{"中", "BINARY", 1, "\xe4"},
+		{"中", "BINARY", 2, "\xe4\xb8"},
+		{"中", "BINARY", 3, "中"},
+		{"中", "BINARY", 5, "中\x00\x00"},
```
Comment (djshow832) [inline expression/builtin_string_test.go]: Test the result when `expr` is a datetime or some types else.

File: expression/evaluator_test.go [hunk 8]
```diff
@@ -136,6 +135,9 @@ func (s *testEvaluatorSuiteBase) datumsToConstants(datums []types.Datum) []Expre
 	constants := make([]Expression, 0, len(datums))
 	for _, d := range datums {
 		ft := s.kindToFieldType(d.Kind())
+		if types.IsNonBinaryStr(&ft) {
+			ft.Collate = d.Collation()
```
Comment (wjhuang2016) [inline expression/evaluator_test.go]: fLen too?

File: expression/builtin_string.go [hunk 9]
```diff
@@ -3651,3 +3654,164 @@ type loadFileFunctionClass struct {
...
+	newSig.length = b.length
+	return newSig
+}
+
+// evalString evals a WEIGHT_STRING(expr [AS (CHAR|BINARY)]) when the expr is non-numeric types.
+// See https://dev.mysql.com/doc/refman/5.7/en/string-functions.html#function_weight-string
+func (b *builtinWeightStringSig) evalString(row chunk.Row) (string, bool, error) {
+	str, isNull, err := b.args[0].EvalString(b.ctx, row)
+	if err != nil {
+		return "", true, err
```
Comment (AilinKid) [inline expression/builtin_string.go]: `return "", false, err` here?

File: expression/builtin_string.go (hunk 6 above)
Comment (bb7133) [reply expression/builtin_string.go]: Yes...it is not used anymore, thanks.

File: expression/builtin_string.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (bb7133) [reply expression/builtin_string.go]: Addressed.

File: expression/builtin_string.go (hunk 9 above)
Comment (bb7133) [reply expression/builtin_string.go]: Thanks.

File: expression/builtin_string_test.go (hunk 7 above)
Comment (bb7133) [reply expression/builtin_string_test.go]: It's kind of complicated to add them here, I've added the integration test and put those cases there, PTAL.

File: expression/evaluator_test.go (hunk 8 above)
Comment (bb7133) [reply expression/evaluator_test.go]: It is not needed anymore.

**Comment by bb7133** [comment]
PTAL @djshow832 @wjhuang2016 @AilinKid

File: expression/integration_test.go [hunk 10]
```diff
@@ -47,6 +48,7 @@ import (
 
 var _ = Suite(&testIntegrationSuite{})
 var _ = Suite(&testIntegrationSuite2{})
+var _ = Suite(&testIntegrationSerialSuite{})
```
Comment (wjhuang2016) [inline expression/integration_test.go]: SerialSuite

**Pull request merged**